File: python/LC_371_SumOfTwoIntegers.py
# https://leetcode.com/problems/sum-of-two-integers
class Solution:
    def getSum(self, a: int, b: int) -> int:
        MASK = 0xFFFFFFF
        MAX_INT = 2**31 -1
        while b != 0:
            sum  = (a ^ b) & MASK
            b =  ((a & b) << 1) & MASK
            a = sum
        return a if a <= MAX_INT else ~(a ^ MASK)
    
sol = Solution()
assert sol.getSum(5, 3) == 8
assert sol.getSum(1, 2) == 3
assert sol.getSum(2, 3) == 5

File: java/src/LC_371_SumOfTwoIntegers.java
// https://leetcode.com/problems/sum-of-two-integers

public class LC_371_SumOfTwoIntegers {

    public void bitwiseSum(int a, int b) {
        while (b != 0) {
            int sum = a ^ b;
            b = (a & b) << 1;
            a = sum;
        }
        System.out.println(a);
    }

    public static void main(String[] args) {
        LC_371_SumOfTwoIntegers solution = new LC_371_SumOfTwoIntegers();
        solution.bitwiseSum(10, 29);
        solution.bitwiseSum(-21, 2);
        solution.bitwiseSum(5, -7);
        solution.bitwiseSum(15, -5);
        solution.bitwiseSum(-5, -6);

    }
}
